pkg/ddl/ddl_api.go

```diff
@@ -1413,6 +1413,17 @@ func getFuncCallDefaultValue(col *table.Column, option *ast.ColumnOption, expr *
 		}
 		return nil, false, dbterror.ErrDefValGeneratedNamedFunctionIsNotAllowed.GenWithStackByArgs(col.Name.String(),
 			fmt.Sprintf("%s with disallowed args", expr.FnName.String()))
+	case ast.JSONObject, ast.JSONArray, ast.JSONQuote: // JSON_OBJECT(), JSON_ARRAY(), JSON_QUOTE()
```

Comment:
Do we need to check these functions in `CreateNewColumn` like `UUIDToBin`?
And please add some UTs like "add column", "modify column" , etc

Reply:
I'm not sure the check in `CreateNewColumn` for `UUIDToBin` etc is actually doing much.

I could allow `ALTER TABLE ... ADD COLUMN ... json DEFAULT (JSON_ARRAY())` etc with this patch:

```patch
diff --git a/pkg/ddl/column.go b/pkg/ddl/column.go
index 16c4a1664b..13c0aff1ae 100644
--- a/pkg/ddl/column.go
+++ b/pkg/ddl/column.go
@@ -1995,7 +1995,10 @@ func generateOriginDefaultValue(col *model.ColumnInfo, ctx sessionctx.Context) (
                oldValue := strings.ToLower(valStr)
                // It's checked in getFuncCallDefaultValue.
                if !strings.Contains(oldValue, fmt.Sprintf("%s(%s(),", ast.DateFormat, ast.Now)) &&
-                       !strings.Contains(oldValue, ast.StrToDate) {
+                       !strings.Contains(oldValue, ast.StrToDate) &&
+                       !strings.Contains(oldValue, ast.JSONArray) &&
+                       !strings.Contains(oldValue, ast.JSONObject) &&
+                       !strings.Contains(oldValue, ast.JSONQuote) {
                        return nil, errors.Trace(dbterror.ErrBinlogUnsafeSystemFunction)
                }
 
```

But let's for now keep this identical to most of the other function defaults and not allow this.

```
mysql> CREATE TABLE t (id INT PRIMARY KEY);
Query OK, 0 rows affected (0.03 sec)

mysql> ALTER TABLE t ADD COLUMN u VARCHAR(36) DEFAULT (UUID());
ERROR 1674 (HY000): Statement is unsafe because it uses a system function that may return a different value on the slave
mysql> ALTER TABLE t ADD COLUMN u BINARY(16) DEFAULT (UUID_TO_BIN(UUID()));
ERROR 1674 (HY000): Statement is unsafe because it uses a system function that may return a different value on the slave
mysql> ALTER TABLE t ADD COLUMN r BIGINT DEFAULT (RAND());
ERROR 1674 (HY000): Statement is unsafe because it uses a system function that may return a different value on the slave
mysql> SELECT TIDB_VERSION()\G
*************************** 1. row ***************************
TIDB_VERSION(): Release Version: v8.2.0-alpha-98-g013a4e6a3f
Edition: Community
Git Commit Hash: 013a4e6a3fa13b7f528f16a43a24133d3de7747b
Git Branch: master
UTC Build Time: 2024-05-07 18:30:53
GoVersion: go1.22.2
Race Enabled: false
Check Table Before Drop: false
Store: unistore
1 row in set (0.01 sec)
```

Note that if we would allow this we need to check how to handle `RAND()` as plain default function and as argument for a JSON function:

```
mysql> CREATE TABLE t (id INT PRIMARY KEY);
Query OK, 0 rows affected (0.03 sec)

mysql> ALTER TABLE t ADD COLUMN j json DEFAULT (JSON_OBJECT('rand', RAND()));
Query OK, 0 rows affected (0.06 sec)

mysql> INSERT INTO t(id) VALUES (1),(2),(3);
Query OK, 3 rows affected (0.00 sec)
Records: 3  Duplicates: 0  Warnings: 0

mysql> TABLE t;
+----+-------------------------------+
| id | j                             |
+----+-------------------------------+
|  1 | {"rand": 0.2941574745757109}  |
|  2 | {"rand": 0.49900682410132774} |
|  3 | {"rand": 0.612561727513151}   |
+----+-------------------------------+
3 rows in set (0.00 sec)
```